specs/FileTypePolicy.md

```diff
@@ -0,0 +1,225 @@
+  /// Adds an event handler for the `FileTypePolicyChecking` event.
+  /// This event will be raised during system FileTypePolicy
+  /// checking the dangerous file extension list.
+  /// 
+  /// Developers can specify their own decision on if allow this file 
+  /// type extension to be saved, or downloaded. Here are two properties
+  /// in `ICoreWebView2FileTypePolicyCheckingEventArgs` to manage the 
+  /// decision, `Cancel` and `SuppressDefaultPolicy`.
+  /// Table of Properties' value and result:
+  /// 
+  /// | Cancel | SuppressDefaultPolicy | Result
+  /// | ------ | ------ | ---------------------  
+  /// | False  | False  | Process to default policy check. It might show 
+  /// |        |        | security warning UI if the file extension is 
+  /// |        |        | dangerous.
+  /// | ------ | ------ | --------------------- 
+  /// | False  | True   | Skip the default policy check and the possible
+  /// |        |        | security warning. Start saving or downloading.
+  /// | ------ | ------ | --------------------- 
+  /// | True   | T or F | Skip the default policy check and the possible
+  /// |        |        | security warning. Abort save or download.
+  HRESULT add_FileTypePolicyChecking(
+      [in] ICoreWebView2StagingFileTypePolicyCheckingEventHandler* eventHandler,
+      [out] EventRegistrationToken* token);
+
+  /// Removes an event handler previously added with `add_FileTypePolicyChecking`.
+  HRESULT remove_FileTypePolicyChecking(
+      [in] EventRegistrationToken token);
+}
+
+
+/// The event args for `FileTypePolicyChecking` event.
+interface ICoreWebView2StagingFileTypePolicyCheckingEventArgs : IUnknown {
+  /// Gets the `Cancel` property.
+  [propget] HRESULT Cancel([out, retval] BOOL* value);
+
+  /// Set if cancel the upcoming save/download. `TRUE` means the action 
+  /// will be cancelled before validations in default policy.
+  /// 
+  /// The default value is `FALSE`.
+  [propput] HRESULT Cancel([in] BOOL value);
+
+  /// Get the extension of file to be saved.
+  ///
+  /// File extension can be empty.
+  ///
+  /// Only final extension will be provided. For example, "*.tar.gz" 
+  /// is a double extension, where the "gz" will be its final extension.
+  ///
+  /// File extension usage in the API is case sensitive.
+  [propget] HRESULT FileExtension([out, retval] LPWSTR* value);
+
+  /// Get the full path of file to be saved. This includes the
+  /// file name and extension.
+  [propget] HRESULT FilePath([out, retval] LPWSTR* value);
+
+  /// Gets the `SuppressDefaultPolicy` property.
+  [propget] HRESULT SuppressDefaultPolicy([out, retval] BOOL* value);
+
+  /// Set if the default policy checking and security warning will be
+  /// suppressed. `TRUE` means it will be suppressed. 
+  /// 
+  /// The default value is `FALSE`.
+  [propput] HRESULT SuppressDefaultPolicy([in] BOOL value);
+
+  /// Get the uri of file source.
+  [propget] HRESULT Uri([out, retval] LPWSTR* value);
```

Comment:
Would be good to be slightly more explicit and say this is the SourceUri

```suggestion
  [propget] HRESULT SourceUri([out, retval] LPWSTR* value);
```

Reply:
We should probably pick a consistent name going forward:
- `PermissionRequested` has `Uri`
- `FrameInfo` has `Source`
- `WebMessageReceived` has `Source`
- `LaunchingExternalUriScheme` has `InitiatingOrigin`
- `ContextMenuTarget` has `SourceUri`
- `ScriptDialogOpening` has `Uri`
- `BasicAuthenticationRequested` has `Uri`

and maybe more.

Reply:
I think SourceUri is probably best since without Uri its just a string named Source and its not necessarily clear that will be a URI and by adding 'Source' it makes it clearer what the URI represents. I've updated the WebView2 API Guidance with a section on this including Andy's list (thx!)